msticpy/analysis/polling_detection.py

```diff
@@ -0,0 +1,170 @@
+
+There is currently only one technique available for filtering polling data which is
+the class PeriodogramPollingDetector.
+"""
+from collections import Counter
+from typing import Tuple
+
+import numpy as np
+import numpy.typing as npt
+
+from scipy import signal, special
+
+from ..common.utility import export
+
+
+@export
+class PeriodogramPollingDetector:
+    """
+    Polling detector using the Periodogram to detect strong frequencies.
+
+    Methods
+    -------
+    detect_polling(timestamps, process_start, process_end, interval)
+        Detect strong periodic frequencies
+
+    """
+
+    def __init__(self) -> None:
+        """Create periodogram polling detector."""
+
+    def _g_test(self, pxx: npt.NDArray, exclude_pi: bool) -> Tuple[float, float]:
+        """
+        Carry out fishers g test for periodicity.
+
+        Fisher's g test tests the null hypothesis that the time series is gaussian white noise
+        against the alternative that there is a deterministic periodic component[1]
+
+        If the length of the time series is even then the intensity at pi should be excluded
+
+        If the length of the power spectral density estimate is larger than 700 then an approximate
+        p value is calculated otherwise the exact p value is calculate.
+
+        This implementation was taken from the R package GeneCycle[2]
+
+        Parameters
+        ----------
+        pxx: ArrayLike
+            Estimate of the power spectral density
+
+        exclude_pi: bool
+            A bool to indicate whether the frequnecy located at pi should be removed.
+
+        Returns
+        -------
+        Tuple[float, float]
+            G test test statistic
+            G test P value
+
+        References
+        ----------
+        [1] M. Ahdesmaki, H. Lahdesmaki and O. Yli-Harja, "Robust Fisher's Test for Periodicity
+        Detection in Noisy Biological Time Series," 2007 IEEE International Workshop on Genomic
+        Signal Processing and Statistics, 2007, pp. 1-4, doi: 10.1109/GENSIPS.2007.4365817.
+        [2] https://github.com/cran/GeneCycle/blob/master/R/fisher.g.test.R
+
+        """
+        if exclude_pi:
+            pxx = pxx[:-1]
+
+        pxx_length = len(pxx)
+        test_statistic = np.max(pxx) / sum(pxx)
+        upper = np.floor(1 / test_statistic).astype("int")
+
+        if pxx_length > 700:
+            p_value = 1 - (1 - np.exp(-pxx_length * test_statistic)) ** pxx_length
+        else:
+            compose = []
+            for j in range(1, upper):
+                compose.append(
+                    (-1) ** (j - 1)
+                    * np.exp(
+                        np.log(special.binom(pxx_length, j))
+                        + (pxx_length - 1) * np.log(1 - j * test_statistic)
+                    )
+                )
+
+            p_value = sum(compose)
+
+        p_value = min(p_value, 1)
+
+        return test_statistic, p_value
+
+    def detect_polling(
```

Comment:
Wondering if it would be good to have a simplified interface that just accepts a DF and does the date->int conversion, grouping, etc.

```python

def detect_polling(
    self,
   data: pd.DataFrame,
   time_column: str,
   groupby: Union[str, List[str]] = None,
   start: Optional[datetime] = None,
   end: Optional[datetime] = None,
   sample_interval: int = 1
):
    if start and end:
        data = data[(data[time_column] >= start) && ((data[time_column] <= end)]
   data["timestamp_int"] = data[time_column].apply(datetime[time_column]).astype("int64")
   p_start = data["timestamp_int"].min()
   p_end = data["timestamp_int"].max()

   if groupby:
       return data.groupby(groupby).apply(
           lambda x: self.detect_polling(x["timestamp_int"], p_start, p_end, sample_interval)
       )
   return self.detect_polling(data["timestamp_int"], p_start, p_end, sample_interval)
```
           
I think this kind of interface might be more easily used by users with data in DFs.
(entirely untested, off the top of my head code)
Perhaps rename the current detect_polling -> detect_polling_arr or something like that?

I guess we need to return the edge/grouping value in the grouped case.

WDYT?

Reply:
Thanks Ian

No problem, I'll change it to accept a DF and then the results will be added as columns to this. The model results would then just be replicated for each edge

#### Non grouped DF
| src_ip | dst_ip | arrival_time | p_value | freq | dominant_interval |
| -------- | -------- | --------------- | ----------- | ----- | ------------------------ |
| 1.1.1.1 | 2.2.2.2 | 2022-01-01 00:00:01 | 0.01 | 0.1 | 10 |
| 1.1.1.1 | 2.2.2.2 | 2022-01-01 00:00:11 | 0.01 | 0.1 | 10 |
| 1.1.1.1 | 2.2.2.2 | 2022-01-01 00:00:21 | 0.01 | 0.1 | 10 |

#### Grouped DF
 | src_ip | dst_ip | arrival_time | p_value | freq | dominant_interval |
| -------- | -------- | --------------- | ----------- | ----- | ------------------------ |
| 1.1.1.1 | 2.2.2.2 | 2022-01-01 00:00:01 | 0.01 | 0.1 | 10 |
| 1.1.1.1 | 2.2.2.2 | 2022-01-01 00:00:11 | 0.01 | 0.1 | 10 |
| 1.1.1.1 | 2.2.2.2 | 2022-01-01 00:00:21 | 0.01 | 0.1 | 10 |
| 3.3.3.3 | 4.4.4.4 | 2022-01-01 00:00:01 | 0.5 | 0.05 | 20 |
| 3.3.3.3 | 4.4.4.4 | 2022-01-01 00:00:51 | 0.5 | 0.05 | 20 |
| 3.3.3.3 | 4.4.4.4 | 2022-01-01 00:01:01 | 0.5 | 0.05 | 20 |

Hopefully that looks like what you were envisioning?

Cheers
Dan

Reply:
Yeah - in fact I was thinking that you could make it more general by using a pandas grouping expression.
If I supply a single column (let's say account_name) it could calculate periodicity of logons for each user.
If I supply a list of columns (2 or more) it would essentially build edges on that column set - e.g. I might have account, host, dest_ip - and it would give periodicity of the tuple of specific account + logged on host + dest IP.
Returning the values columns of the grouping in the DF as you show above. 
I'm sure you get what I am talking about - just wanted to make sure that you weren't restricting it to 2 columns.

Reply:
Yeah we're on the same page :)
The src --> dst pair above was just an example. I think I was confusing matters by getting hung up on edges.

Like in your example above there will be a groupby parameter that the analyst can use. If groupby is `None` it's assumed that all of the timestamps in the dataframe are for a single group/edge/entity

Reply:
Hey @ianhelle 
I've made the changes and resubmitted my pull request
I decided to remove the start and end parameters as I felt it was adding too much complexity, such as
* How to handle the different possible datetime representations
* If groupby is passed then start and ends could either be single values or a list of values for each group
So by default it will use the min and max timestamp for each edge.

What do you think about this?